File: ml-benchmark/autolifter/single-pass/count10s2.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 5;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

count10s2 = fix (
  \f: Bool -> List -> Int. \s0: Bool. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let upd = if and s0 (== h 2) then 1 else 0 in
    let s0 = or (== h 1) (and s0 (== h 0)) in
      + upd (f s0 t)
  end
) false;

/*first_non_zero = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> if == h 0 then f t else h
  end
);*/

main = single_pass count10s2;

File: ml-benchmark/autolifter/single-pass/count10s2.ml
type list_ = Cons of int * list_ | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let rec count10s2_aux s0 xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) ->
      let upd = if s0 && (h = 2) then 1 else 0 in
      let s0 =
        let b1 = h = 1 in
        let b2 = s0 && (h = 0) in
        b1 || b2
      in
      upd + count10s2_aux s0 t

let count10s2 xs = count10s2_aux false xs

let program = single_pass count10s2